### DISCUSSION TITLE
Source Generators

### DISCUSSION THREAD

**Discussion by Jure-BB** [jamescourtney/FlatSharp#297] [Ideas]
Would it be possible to replace runtime compilation with source generators? 

Doing pre-compilation when using .fbs files and runtime compilation when using attributes feels a bit inconsistent. Moving to source generators when using attributes would unify the library in sense that all generation happens at compile time. 
In addition to making it more consistent it would also be a slight improvement to startup time and smaller binary size.

**Comment by jamescourtney** [comment]
This is a good idea. Thank you for suggesting it. The backstory here (not that you asked) is that the reflection-based attribute system came first, and the FlatSharp compiler is layered on top of that. I mostly try to nudge people towards using FBS files instead of attributes these days since there is less to mess up and you get better interop.

Source generators are relatively new to me. It looks like Unity supports them now, which is neat. I need to understand if legacy MSBuild will work with them as well (probably not), and if I care :) I would be happy to be out out of the business of having a custom targets file.

I'm going to add this to the list in #276. I usually do major versions by waiting for a critical mass of breaking changes to accumulate and then do all of those in one shot.

**Reply by Jure-BB** [reply]
Doing it in a major version with braking changes sounds like a good plan. Since pre-compilation is already possible with .fbs files, this is more of a nice-to-have feature, and not all that crucial. 

Thanks for informative reply. You made a very nice library.

**Comment by Jure-BB** [comment]
@jamescourtney I'm excited to see this has been now implemented in v7. Congratulations for major new release!

**Reply by jamescourtney** [reply]
Thanks for the kind words! I'm not quite there yet with source generators, but deprecating reflection-only mode is one step along that path. Still using the custom targets file and FlatSharp.Compiler.dll for now. Ultimately, I do want to move to source generators, but it will take some time (and a bunch of refactoring) to get there.

Be sure and let me know if you run into any issues with V7. Happy to help how I can.

**Reply by Jure-BB** [reply]
I see. So, defining structs and classes to be serializable by annotating them with attributes is not supported in v7 anymore. When I saw that support for runtime code generation is being dropped, I mistakenly assumed it is being replaced by source generators. It seem v7 is a quality release nonetheless. 

I'm not too familiar with implementing source generators, but if I'm not mistaken they can't be used for generating C# code from FSB files. But I'm not sure. 
 
[MemoryPack](https://github.com/Cysharp/MemoryPack) serializer is similar to FlatBuffers and uses source generators, so it might be a good reference on how to use source generators for this kind of task. However, it doesn't use schema files for defining serializable types, only attribute annotations.

**Reply by jamescourtney** [reply]
I still do plan to move towards source generators, but it's a lot of work and will take some time. Deprecating runtime mode is a prerequisite for that work, though. Next means moving all ~900 test cases to use precompiled mode instead of reflection mode, then rewriting the core of FlatSharp. So, it's going to be a lot of work and won't be done any time soon :)
